ADBDEV-4861 jit: Support opaque pointers in LLVM 16.

dnskvlnk opened the pull request:
This is a backport of https://github.com/arenadata/gpdb/commit/15ddc9725eb73d97a16652c7c90d993302773544 from PostgreSQL

Remove use of LLVMGetElementType() and provide the type of all pointers
to LLVMBuildXXX() functions when emitting IR, as required by modern LLVM
versions[1].

 * For LLVM <= 14, we'll still use the old LLVMBuildXXX() functions.
 * For LLVM == 15, we'll continue to do the same, explicitly opting
   out of opaque pointer mode.
 * For LLVM >= 16, we'll use the new LLVMBuildXXX2() functions that take
   the extra type argument.

The difference is hidden behind some new IR emitting wrapper functions
l_load(), l_gep(), l_call() etc.  The change is mostly mechanical,
except that at each site the correct type had to be provided.

In some places we needed to do some extra work to get functions types,
including some new wrappers for C++ APIs that are not yet exposed by in
LLVM's C API, and some new "example" functions in llvmjit_types.c
because it's no longer possible to start from the function pointer type
and ask for the function type.

Back-patch to 12, because it's a little tricker in 11 and we agreed not
to put the latest LLVM support into the upcoming final release of 11.

[1] https://llvm.org/docs/OpaquePointers.html

Reviewed-by: Dmitry Dolgov <9erthalion6@gmail.com>
Reviewed-by: Ronan Dunklau <ronan.dunklau@aiven.io>
Reviewed-by: Andres Freund <andres@anarazel.de>
Discussion: https://postgr.es/m/CA%2BhUKGKNX_%3Df%2B1C4r06WETKTq0G4Z_7q4L4Fxn5WWpMycDj9Fw%40mail.gmail.com

Conflicts:
    src/backend/jit/llvm/llvmjit_expr.c

Postgres doesn't have v_parent variable which is gpdb specific, and it
caused a conflict. This commit resolves it leaving v_parent as is.

macdice committed:
jit: Support opaque pointers in LLVM 16.
This is a backport of 15ddc97 from PostgreSQL

Remove use of LLVMGetElementType() and provide the type of all pointers
to LLVMBuildXXX() functions when emitting IR, as required by modern LLVM
versions[1].

 * For LLVM <= 14, we'll still use the old LLVMBuildXXX() functions.
 * For LLVM == 15, we'll continue to do the same, explicitly opting
   out of opaque pointer mode.
 * For LLVM >= 16, we'll use the new LLVMBuildXXX2() functions that take
   the extra type argument.

The difference is hidden behind some new IR emitting wrapper functions
l_load(), l_gep(), l_call() etc.  The change is mostly mechanical,
except that at each site the correct type had to be provided.

In some places we needed to do some extra work to get functions types,
including some new wrappers for C++ APIs that are not yet exposed by in
LLVM's C API, and some new "example" functions in llvmjit_types.c
because it's no longer possible to start from the function pointer type
and ask for the function type.

Back-patch to 12, because it's a little tricker in 11 and we agreed not
to put the latest LLVM support into the upcoming final release of 11.

[1] https://llvm.org/docs/OpaquePointers.html

Reviewed-by: Dmitry Dolgov <9erthalion6@gmail.com>
Reviewed-by: Ronan Dunklau <ronan.dunklau@aiven.io>
Reviewed-by: Andres Freund <andres@anarazel.de>
Discussion: https://postgr.es/m/CA%2BhUKGKNX_%3Df%2B1C4r06WETKTq0G4Z_7q4L4Fxn5WWpMycDj9Fw%40mail.gmail.com

Conflicts:
    src/backend/jit/llvm/llvmjit_expr.c

PostgreSQL doesn't have v_parent variable which is GPDB specific, and it
caused a conflict. This commit resolves it leaving v_parent as is.

BenderArenadata commented:
Failed job Build 7 for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1215512

BenderArenadata commented:
Failed job Build 7 for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1221860

BenderArenadata commented:
Failed job Build 7 for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1232479

BenderArenadata commented:
Failed job Build 7 for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1234982

Commit:
Fix JIT support for GPDB

This commit updates v_parent assignment with new l_load_struct_gep interface
providing type reference and fixes various nodes that aren't represented
in PostgreSQL.
Also, this commit fixes EEOP_AGG_PLAIN_PERGROUP_NULLCHECK operator that
generated invalid LLVM IR passing struct reference to getelementptr instead of
a regular pointer which resulted in SIGILL error.

Commit:
Update LLVM and describe how to start JIT tests in readme

RekGRpth commented:
backport should be cherry-pick to save authorship
also you need explicit point, what was changed and why

red1452 commented on arenadata/readme.md:
```diff
@@ -32,6 +32,27 @@ We need to execute [../concourse/scripts/ic_gpdb.bash](../concourse/scripts/ic_g
 
 To use gdb inside the container, add the `--privileged` flag to the run command.
```
This line and two above are for all regression tests. Move it under new regression tests.

dnskvlnk replied:
Done

red1452 commented on arenadata/readme.md:
````diff
@@ -32,6 +32,27 @@ We need to execute [../concourse/scripts/ic_gpdb.bash](../concourse/scripts/ic_g
+
+* optimizer=on
+```bash
+ docker run --name gpdb7_opt_on --rm -it -e TEST_OS=centos \
````
Target OS for GPDB 7 is rocky8. Maybe, it should be changed here.

dnskvlnk replied:
Done

red1452 commented on arenadata/Dockerfile:
```diff
@@ -24,15 +24,14 @@ RUN dnf makecache && \
+                llvm-devel \
+                clang-resource-filesystem \
+                clang-libs \
+                clang && \
```
Maybe print it into one line?

dnskvlnk replied:
It has always been that way so I don't think we should change it

red1452 reviewed (commented):
Add backport info to the first commit with short info about conflicts.

red1452 commented on arenadata/Dockerfile:
```diff
@@ -24,15 +24,14 @@ RUN dnf makecache && \
+                llvm-devel \
+                clang-resource-filesystem \
+                clang-libs \
+                clang && \
```
`dnf -y install llvm-devel clang` is enough. All other packages will be installed as dependencies.

dnskvlnk replied:
Done

KnightMurloc commented on arenadata/readme.md:
````diff
@@ -27,6 +27,27 @@ We need to execute [../concourse/scripts/ic_gpdb.bash](../concourse/scripts/ic_g
+* optimizer=on
+```bash
+ docker run --name gpdb7_opt_on --rm -it -e TEST_OS=centos \
+  -e MAKE_TEST_COMMAND="-k PGOPTIONS='-c optimizer=on -c jit=on -c jit_above_cost=0 -c optimizer_jit_above_cost=0 -c gp_explain_jit=off' installcheck" \
````
why not `installcheck-world`?

dnskvlnk replied:
Bc upstream launches it that way

KnightMurloc reviewed (commented):
I think we can use rocky linux 8.9 as the base image (or just rockylinux:8) since the reason for using the earlier version was the need for llvm 15.

KnightMurloc commented:
I think it's worth combining the commits of the backport and the gpdb-specific fixes (so that there are 2 commits left) since these commits have no value individually. I'm also not sure if it's worth mentioning the problem with `EEOP_AGG_PLAIN_PERGROUP_NULLCHECK` since it doesn't show up before or after the patch.

dnskvlnk commented:
@KnightMurloc We discussed this and decided to keep backport and fix commits separate bc backport's changes must be untouched and it must be clear what changes were made. EEOP_AGG_PLAIN_PERGROUP_NULLCHECK was mentioned bc the second commit fixes it

KnightMurloc commented:
squash the last two commits.

red1452 commented on arenadata/Dockerfile:
```diff
@@ -23,16 +23,7 @@ RUN dnf makecache && \
-                clang-resource-filesystem-15.0.7-1.module+el8.8.0+1144+0a4e73bd \
-                clang-libs-15.0.7-1.module+el8.8.0+1144+0a4e73bd \
-                clang-15.0.7-1.module+el8.8.0+1144+0a4e73bd && \
+RUN dnf -y install llvm-devel clang && \
```
I think, we can move this execution of `dnf` to the previous RUN step.

dnskvlnk replied:
Done

red1452 reviewed (commented):
Change description and commits' messages: gpdb -> GPDB, jit -> JIT, llvm -> LLVM, postgres -> PostgreSQL.

KnightMurloc commented on arenadata/Dockerfile:
```diff
@@ -18,23 +18,12 @@ RUN dnf makecache && \
     dnf -y install libicu perl-ExtUtils-Embed perl-Env perl-JSON && \
     dnf -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel && \
     dnf -y install python39-psycopg2 python39-psutil python39-pyyaml python39-lxml && \
+	dnf -y install llvm-devel clang && \
```
use spaces.
```suggestion
    dnf -y install llvm-devel clang && \
```

dnskvlnk replied:
Done

Commit:
Merge branch 'adb-7.1.0' into ADBDEV-4861

andr-sokolov commented on arenadata/Dockerfile:
```diff
@@ -18,23 +18,12 @@ RUN dnf makecache && \
     dnf -y install libicu perl-ExtUtils-Embed perl-Env perl-JSON && \
     dnf -y install perl-IPC-Run perl-Test-Base libxslt-devel openldap-devel && \
     dnf -y install python39-psycopg2 python39-psutil python39-pyyaml python39-lxml && \
+    dnf -y install llvm-devel clang && \
```
Is it necessary to run dnf one more time? Can we add these two packages to the existing package list, for example in the 14th line? I think the fewer times we run dnf, the faster the installation is.

dnskvlnk replied:
I think it would be justified if we had one dnf run for all the packages but it's called multiple times so I think it's better to leave clang and llvm on a separate line to show that they're connected and keep diff obvious so one would see what changes were made at a glance. It wouldn't be so clear if you left it on the same line with python
